includes/reader-revenue/woocommerce/class-woocommerce-cli.php

```diff
@@ -162,6 +162,188 @@ function( $subscription ) {
+	 * @param array $assoc_args Associative args.
+	 */
+	public function fix_missing_next_payment_dates( $args, $assoc_args ) {
+		$dry_run    = ! empty( $assoc_args['dry-run'] );
+		$now        = time();
+		$batch_size = ! empty( $assoc_args['batch-size'] ) ? intval( $assoc_args['batch-size'] ) : 50;
+		$start_date = ! empty( $assoc_args['start-date'] ) ? strtotime( $assoc_args['start-date'] ) : strtotime( '-90 days', $now );
+
+		if ( ! $dry_run ) {
+			\WP_CLI::line( "\n=====================\n=     LIVE MODE     =\n=====================\n" );
+		} else {
+			\WP_CLI::line( "\n===================\n=     DRY RUN     =\n===================\n" );
+		}
+		sleep( 2 );
+
+		\WP_CLI::log(
+			'
+Fetching active subscriptions with missing or missed next_payment dates...
+		'
+		);
+
+		$query_args    = [
+			'subscriptions_per_page' => $batch_size,
+			'subscription_status'    => [ 'active', 'pending' ],
+			'offset'                 => 0,
+		];
+		$processed     = 0;
+		$subscriptions = \wcs_get_subscriptions( $query_args );
+		$total_revenue = 0;
+		$results       = [];
+
+		while ( ! empty( $subscriptions ) ) {
+			foreach ( $subscriptions as $subscription_id => $subscription ) {
+				array_shift( $subscriptions );
+
+				// If the subscription start date is before the $args start date, we're done.
+				if ( strtotime( $subscription->get_date( 'start' ) ) < $start_date ) {
+					$subscriptions = [];
+					break;
+				}
+
+				$result = self::calculate_next_payment_date( $subscription, $dry_run );
+				if ( ! $result ) {
+					continue;
+				}
+
+				if ( $result['missed_periods'] ) {
+					$total_revenue += $result['missed_total'];
+				}
+
+				$results[] = $result;
+				$processed++;
+
+				// Get the next batch.
+				if ( empty( $subscriptions ) ) {
+					$query_args['offset'] += $batch_size;
+					$subscriptions        = \wcs_get_subscriptions( $query_args );
+				}
+			}
+		}
+
+		if ( empty( $results ) ) {
+			\WP_CLI::log( 'No subscriptions with missing next_payment dates found in the given time period.' );
+		} else {
+			\WP_CLI\Utils\format_items(
+				'table',
+				$results,
+				[
+					'ID',
+					'status',
+					'start_date',
+					'next_payment_date',
+					'end_date',
+					'billing_period',
+					'missed_periods',
+					'missed_total',
+				]
+			);
+			\WP_CLI::success(
+				sprintf(
+					'Finished processing %d subscriptions. %s',
+					$processed,
+					$total_revenue ? 'Total missed revenue: ' . \wp_strip_all_tags( html_entity_decode( \wc_price( $total_revenue ) ) ) : ''
+				)
+			);
+		}
+		\WP_CLI::line( '' );
+	}
+
+	/**
+	 * Given a subscription, calculates the next payment date and missed payments.
+	 *
+	 * @param WC_Subscription $subscription The subscription.
+	 * @param bool            $dry_run If set, will not make any changes.
+	 *
+	 * @return array|false The result array or false if the subscription is broken.
+	 */
+	public static function calculate_next_payment_date( $subscription, $dry_run = false ) {
```

Comment:
Nit: the function name suggests the the return value is a date. Could be something along the lines of `validate_subscription_dates`.

Reply:
Updated to `validate_subscription_dates` in 879ba1a